src/Microsoft.AspNetCore.Mvc.Versioning/Versioning/ApiVersionCollator.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace Microsoft.AspNetCore.Mvc.Versioning
+{
+    using Microsoft.AspNetCore.Mvc.Abstractions;
+    using Microsoft.AspNetCore.Mvc.Controllers;
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Contracts;
+    using System.Linq;
+
+    /// <summary>
+    /// Represents an object that collates <see cref="ApiVersion">API versions</see> per <see cref="ActionDescriptor">action</see>.
+    /// </summary>
+    [CLSCompliant( false )]
+    public class ApiVersionCollator : IActionDescriptorProvider
+    {
+        /// <inheritdoc />
+        public int Order { get; protected set; }
+
+        /// <inheritdoc />
+        public virtual void OnProvidersExecuted( ActionDescriptorProviderContext context )
+        {
+            foreach ( var actions in GroupActionsByController( context.Results ) )
+            {
+                var collatedModel = CollateModel( actions );
+
+                foreach ( var action in actions )
+                {
+                    var model = action.GetProperty<ApiVersionModel>();
+
+                    if ( model != null )
+                    {
+                        action.SetProperty( model.Aggregate( collatedModel ) );
```

Comment:
Is this guaranteed to be thread safe? I can see that Microsoft.AspNetCore.Mvc.Abstractions.ActionDescriptor the Properties "property" is backed by a simple Dictionary but I'm unclear of the flow here and if there is a case where there can be multiple threads accessing this (which is the problem you are trying to solve with this PR)